pkg/tsdb/azuremonitor/utils/error_parsing.go

```diff
@@ -0,0 +1,69 @@
+package utils
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+)
+
+type ErrorObject struct {
+	Code                    string        `json:"code"`
+	Message                 string        `json:"message"`
+	Line                    int           `json:"line,omitempty"`
+	CharacterPositionInLine int           `json:"characterPositionInLine,omitempty"`
+	Token                   string        `json:"token,omitempty"`
+	ExpectedToken           string        `json:"expectedToken,omitempty"`
+	Details                 []ErrorObject `json:"details"`
+}
+type errorResponse struct {
+	ErrorObject `json:"error"`
+}
+
+type XModel errorResponse
+type XModelExceptions struct {
+	XModel
+	Other *string // Other won't raise an error
+}
+
+func (er *errorResponse) UnmarshalJSON(data []byte) error {
+	var me XModelExceptions
+	dec := json.NewDecoder(bytes.NewReader(data))
+	dec.DisallowUnknownFields() // Force errors
+
+	if err := dec.Decode(&me); err != nil {
+		return err
+	}
+
+	*er = errorResponse(me.XModel)
+	return nil
+}
+
+func ParseError(e []byte) (string, error) {
+	er := errorResponse{}
+	err := er.UnmarshalJSON(e)
+	if err != nil {
+		return string(e), err
+	}
+
+	errString := er.Code + "\n"
+	for _, d := range er.Details {
+		errString += d.Code + ": " + d.Message
+		if d.Line != 0 {
+			errString += " at line " + fmt.Sprint(d.Line)
+		}
+		if d.CharacterPositionInLine != 0 {
```

Comment:
`code` is the type identifier of an error, e.g. `BadRequest`, `InvalidQuery`, `UnknownFunction` etc.

Custom error fields (which are not `code`, `message`, `details` etc.) are specific to a particular error `code`.
Therefore, you could check for the error `code` first:
```go
if d.Code == "ParserFailure" {
  // since it's ParserFailure expect Token, ExpectedToken, CharacterPositionInLine, Line etc...
}
```

Reply:
Make sense, 
But also it does not seem very generalisable... 
I'm having a look at a more generic frontend solution, to display as a json rather than try and parse in the backend, see: https://github.com/grafana/grafana/pull/44877